R/tree.R

```diff
@@ -0,0 +1,314 @@
+#'
+#' These methods control how the value of a given node is printed. New methods
+#' can be added if support is needed for a novel class
+#'
+#' @inheritParams tree
+#'
+#' @export
+tree_label <- function(x, val_printer, class_printer, remove_newlines){
+  UseMethod("tree_label")
+}
+
+#' @export
+tree_label.function <- function(x, ...){
+  crayon::italic("function(){...}")
+}
+
+#' @export
+tree_label.environment <- function(x,...){
+  format.default(x)
+}
+
+#' @export
+tree_label.NULL <- function(x,...){
+  "<NULL>"
+}
+
+#' @export
+tree_label.character <- function(x, remove_newlines, ...){
+
+  # Get rid of new-line so they don't break tree flow
+  if(remove_newlines){
+    x <- gsub("\\n", replacement = " ", x = x, perl = TRUE)
+  }
+
+  # Shorten strings if needed
+  max_standalone_length <- 35
+  max_vec_length <- 15
+  max_length <- if(length(x) == 1) max_standalone_length else max_vec_length
+  x <- ifelse(
+    nchar(x) > max_length,
+    # Since we add an elipses we need to take a bit more than the max length
+    # off. The gsub adds elipses but also makes sure we dont awkwardly end on
+    # a space.
+    gsub(x = substr(x, start = 1, max_length - 3),
+         pattern = "\\s*$",
+         replacement = "...",
+         perl = TRUE),
+    x
+  )
+
+  tree_label.default(paste0("\"", x, "\""),...)
+}
+
+
+
+#' @export
+tree_label.default <- function(x, val_printer, class_printer, remove_newlines){
+
+  # There are a few psuedo-types that we want different printing behavior for.
+  # Since s3 methods cant differentiate between something like a single
+  # character and a vector of characters we use some logical branching here to
+  # try and use the best printing type for the passed value.
+  is_atomic_value <- is.atomic(x)
+  is_environment <- is.environment(x)
+
+  if(is_atomic_value) {
+
+    num_els <- length(x)
+    if(num_els > 1) {
+      # Atomic vectors are truncated to a max of 10 elements and printed inline
+      x <- as.character(x)
+      if(num_els > 10){
+        x <- head(x, 10)
+        x <- c(x, paste0("...(n = ", num_els, ")"))
+      }
+      x <- paste(x, collapse = ",")
+    }
+
+    # Single length atomics just go through unscathed
+    val_printer(x)
+
+  } else if(is.function(x)) {
+    # Lots of times function-like functions don't actually trigger the s3 method
+    # for function because they dont have function in their class-list. This
+    # catches those.
+    tree_label.function(x, class_printer, val_printer)
+  } else if(is.environment(x)) {
+    # Environments also tend to have the same trouble as functions. For instance
+    # the srcobject attached to a function's attributes is an environment but
+    # doesn't report as one to s3.
+    tree_label.environment(x)
+  } else {
+    # The "base-case" is simply a list-like object. Here we use curly braces if
+    # it has named elements and print the class name
+    delims <- if(!is.null(names(x))) c("{","}") else c("[", "]")
```

Comment:
To follow our existing conventions I think this needs to use `<>` as delimiters, and we'd need some other way of distinguishing named vs unnamed lists.

Reply:
Switched to the `<>` delimiters. Also switched to the `waldo` style `oo_type<class/and/inhertance>` printing.
As for the naming, I think the fact that we print the names for a child on its branch is probably good enough.